torchvision/models/video/s3d.py

```diff
@@ -0,0 +1,173 @@
+        )
+        self.branch2 = nn.Sequential(
+            Conv3dNormActivation(
+                in_planes, b2_mid, kernel_size=1, stride=1, norm_layer=partial(nn.BatchNorm3d, eps=1e-3, momentum=1e-3)
+            ),
+            SepConv3d(b2_mid, b2_out, kernel_size=3, stride=1, padding=1),
+        )
+        self.branch3 = nn.Sequential(
+            nn.MaxPool3d(kernel_size=(3, 3, 3), stride=1, padding=1),
+            Conv3dNormActivation(
+                in_planes, b3_out, kernel_size=1, stride=1, norm_layer=partial(nn.BatchNorm3d, eps=1e-3, momentum=1e-3)
+            ),
+        )
+
+    def forward(self, x):
+        x0 = self.branch0(x)
+        x1 = self.branch1(x)
+        x2 = self.branch2(x)
+        x3 = self.branch3(x)
+        out = torch.cat((x0, x1, x2, x3), 1)
+
+        return out
+
+
+class S3D(nn.Module):
+    """S3D main class.
+
+    Args:
+        num_class (int): number of classes for the classification task.
+
+    Inputs:
+        x (Tensor): batch of videos with dimensions (batch, channel, time, height, width)
+    """
+
+    def __init__(
+        self,
+        num_classes: int = 400,
+    ) -> None:
+        super().__init__()
+        self.base = nn.Sequential(
+            SepConv3d(3, 64, kernel_size=7, stride=2, padding=3),
+            nn.MaxPool3d(kernel_size=(1, 3, 3), stride=(1, 2, 2), padding=(0, 1, 1)),
+            Conv3dNormActivation(
+                64, 64, kernel_size=1, stride=1, norm_layer=partial(nn.BatchNorm3d, eps=1e-3, momentum=1e-3)
+            ),
+            SepConv3d(64, 192, kernel_size=3, stride=1, padding=1),
+            nn.MaxPool3d(kernel_size=(1, 3, 3), stride=(1, 2, 2), padding=(0, 1, 1)),
+            SepInceptionBlock3D(192, [[64], [96, 128], [16, 32], [32]]),
+            SepInceptionBlock3D(256, [[128], [128, 192], [32, 96], [64]]),
+            nn.MaxPool3d(kernel_size=(3, 3, 3), stride=(2, 2, 2), padding=(1, 1, 1)),
+            SepInceptionBlock3D(480, [[192], [96, 208], [16, 48], [64]]),
+            SepInceptionBlock3D(512, [[160], [112, 224], [24, 64], [64]]),
+            SepInceptionBlock3D(512, [[128], [128, 256], [24, 64], [64]]),
+            SepInceptionBlock3D(512, [[112], [144, 288], [32, 64], [64]]),
+            SepInceptionBlock3D(528, [[256], [160, 320], [32, 128], [128]]),
+            nn.MaxPool3d(kernel_size=(2, 2, 2), stride=(2, 2, 2), padding=(0, 0, 0)),
+            SepInceptionBlock3D(832, [[256], [160, 320], [32, 128], [128]]),
+            SepInceptionBlock3D(832, [[384], [192, 384], [48, 128], [128]]),
+        )
+        self.fc = nn.Conv3d(1024, num_classes, kernel_size=1, stride=1, bias=True)
+
+    def forward(self, x):
+        y = self.base(x)
+        y = F.avg_pool3d(y, (2, y.size(3), y.size(4)), stride=1)
+        y = self.fc(y)
+        y = y.view(y.size(0), y.size(1), y.size(2))
+        logits = torch.mean(y, 2)
+
+        return logits
+
+
+class S3D_Weights(WeightsEnum):
+    pass
+
+
+@register_model()
```

Comment:
Once you make any necessary change on the architecture, you would need to produce expected files so that all the tests work. This is very easy, you basically run the test for the specific model one by passing a specific flag:
```
EXPECTTEST_ACCEPT=1 pytest test/test_models.py -k s3d
```

You can read more at #5319

Reply:
Noob Q: Are the expected files just the tensor outputs for running the forward pass? (like [here](https://github.com/pytorch/vision/blob/main/test/test_models.py#L622-L623)) 

I wonder how does this testing strategy handle the following cases:
- The forward signature contains optional (e.g., attention masks) or boolean args which can trigger different code paths / logic
- The output of the forward pass is compounded, e.g., a namedtuple with multiple fields of different types like [here](https://github.com/facebookresearch/multimodal/blob/main/torchmultimodal/models/gpt.py#L18-L22)
- Error raising statements

The usual way is to write test cases manually for each of the codepaths within the same model. Looks like TorchVision just focuses on the default codepath?

Reply:
Also noticed that you set random seed just once before calling the model? [here](https://github.com/pytorch/vision/blob/main/test/test_models.py#L606)

If at some point the model gets refactored, e.g., by changing the order of the layers containing randomness, it will break the test due to change in the sequence of random numbers?

Reply:
@langong347 Good points. Here are some thoughts:

- This testing infra discussed here only covers the forward for every task. Typically all tasks receive the same inputs, so we have separate tests for them. 
- If you have a set of masked models, you would create tests for all of these as well. 
- In our case, we have also various [other tests](https://github.com/pytorch/vision/blob/main/test/test_models.py#L411-L571) to cover other codepaths, including testing for specific [error raising](https://github.com/pytorch/vision/blob/main/test/test_models.py#L800-L825) statements.

> If at some point the model gets refactored, e.g., by changing the order of the layers containing randomness, it will break the test due to change in the sequence of random numbers?

Correct. If the order gets changed, then the developer would have to update the expected file and explain on the PR why that is the case. I agree that this is problematic and our plan is to refactor our tests to load actual weights going forwards. This will give us more stable results (especially for Object Detection models), test the loading of the weights and fix other similar corner-cases.

Reply:
@datumbox Thanks! I notice your pattern of having one type of test for a class of similar models. Wonder how TorchVision tests have evolved to this high degree of uniformity? Is there a test design doc or wiki that I may take a look?